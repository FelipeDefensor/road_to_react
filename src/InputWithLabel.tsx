import * as React from 'react'
import styles from "./App.module.css"

type InputWithLabelProps = {
  id: string,
  value: string,
  isFocused?: boolean,
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void,
  type?: string
  children: React.ReactNode
  }

export const InputWithLabel: React.FC<InputWithLabelProps> = ({
   id,
   value,
   isFocused,
   onInputChange,
   type='text',
   children
  }) => {

  return (
    <div>
      <label htmlFor={id} className={styles.label}>{children}</label>
      &nbsp;
      <input
        type={type}
        id={id}
        value={value}
        autoFocus={isFocused}
        onChange={onInputChange}
        className={styles.input}
      />
    </div>
  )
}