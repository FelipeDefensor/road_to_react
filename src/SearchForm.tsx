import * as React from 'react'
import styles from "./App.module.css"
import { InputWithLabel } from './InputWithLabel'

type SearchParams = {
    searchTerm: string,
    onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void,
    onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void,
    buttonClass: string
  }
  
  const SearchForm: React.FC<SearchParams> = ({
    searchTerm,
    onSearchSubmit,
    onSearchInput,
    buttonClass,
  }: SearchParams) => (
    <form onSubmit={onSearchSubmit} className={styles.searchForm}>
      <InputWithLabel
        id='search'
        onInputChange={onSearchInput}
        isFocused
        value={searchTerm}> 
        <strong>Search:</strong>
      </InputWithLabel>
      <button type="submit" disabled={!searchTerm} className={`${styles.button} ${buttonClass}`}>
          Submit
      </button>
    </form>
  )

  export { SearchForm }