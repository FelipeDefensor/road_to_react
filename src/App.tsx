import * as React from 'react'
import axios from 'axios'
import styles from "./App.module.css"

import { List } from "./List"
import { SearchForm } from "./SearchForm"

const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query='

export enum Action {
  STORIES_FETCH_INIT,
  STORIES_FETCH_SUCCESS,
  STORIES_FETCH_FAILURE,
  LOAD_PREVIOUS_STORIES,
  REMOVE_STORY
}

export type Story = {
  title: string
  url: string
  author: string
  num_comments: number,
  points: number,
  objectID: number
}

export type ReducerAction = {
  type: Action
  payload?: any
}

type StoryState = {
  data: Story[]
  isError: boolean
  isLoading: boolean
}

const storiesReducer = (state:StoryState, action: ReducerAction) => {
  switch (action.type) {
    case Action.STORIES_FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: false
      }
    case Action.STORIES_FETCH_SUCCESS:
      return {
        ...state,
        data: action.payload,
        isLoading: false,
        isError: false
      }
    case Action.LOAD_PREVIOUS_STORIES:
        console.log(action)
        return {
          ...state,
          data: action.payload,
          isLoading: false,
          isError: false
      }
    case Action.STORIES_FETCH_FAILURE:
      return {
        ...state,
        isLoading: false,
        isError: true
      }
    case Action.REMOVE_STORY:
      return {
        ...state,
        data: state.data.filter((story) => story.objectID !== action.payload.objectID),
        isLoading: false,
        isError: false
        }
    default:
      throw new Error('Unrecognized action.')
  }
}

const getSumComments = (stories: Story[]) => {
  return stories.reduce((result, value) => result + value.num_comments, 0
  )
}

const App = () => {
  const useStorageState = (
    key: string,
    initialState: string
    ): [string, (newValue: string) => void] => {
    const isMounted = React.useRef(false)

    console.log('------')
    console.log('Rendering App')

    const [value, setValue] = React.useState(
      localStorage.getItem(key) || initialState
    )

    React.useEffect(() => {
      if (!isMounted.current) {
        isMounted.current = true
      } else {
        localStorage.setItem(key, value)
      }
    }, [value, key])

    return [value, setValue]
  }

  type PreviousSearch = {
    term: string,
    result: StoryState
  }

  const [searchTerm, setSearchTerm] = useStorageState('search', 'React')
  const [lastSearches, setLastSearchTerms] = React.useState<PreviousSearch[]>([])
  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`)
  const [stories, dispatchStories] = React.useReducer(
    storiesReducer,
    { data: [], isLoading: false, isError: false}
  )

  function handleSearchSubmit (event: React.FormEvent<HTMLFormElement>) {
    setUrl(`${API_ENDPOINT}${searchTerm}`)

    console.log(`searchTerm=${ searchTerm }`)

    event.preventDefault()
    
  }

  function handleSearchInput (event: React.ChangeEvent<HTMLInputElement>) {
    console.log(event)
    setSearchTerm(event.target.value)
  }
  

  const handleFetchStories = React.useCallback(async () => {
    dispatchStories({ type: Action.STORIES_FETCH_INIT})

    try {
      const result = await axios.get(url)
    
      dispatchStories({
          type: Action.STORIES_FETCH_SUCCESS,
          payload: result.data.hits
      })

      const newSearchTerms = [...lastSearches]
      newSearchTerms.unshift({
        term: searchTerm,
        result: result.data.hits
      })

      if (newSearchTerms.length === 6) {
        newSearchTerms.pop()
      }

      setLastSearchTerms(newSearchTerms)
      console.log(lastSearches)


    } catch {
      dispatchStories ({ type: Action.STORIES_FETCH_FAILURE})
    }}, [url])
    
  React.useEffect(() => { handleFetchStories() }, [handleFetchStories])

  function onLoadPreviousSearch (term, result) {
    setSearchTerm(term)
    dispatchStories({
      type: Action.LOAD_PREVIOUS_STORIES,
      payload: result
    })
  }

  const PreviousSearchButton = ({ term, result }) => {
    return <button onClick={() => onLoadPreviousSearch(term, result)}>{term}</button>
  }

  return (
    <div className={styles.container}>
      <h1 className={styles.headlinePrimary}>My Hacker Stories with {getSumComments(stories.data)} comments.</h1>
      {stories.isLoading ? <p>Loading...</p> : ''}
      {stories.isError ? <p>ERROR LOADING DATA</p> : ''}
      <SearchForm
        searchTerm={searchTerm}
        onSearchSubmit={handleSearchSubmit}
        onSearchInput={handleSearchInput}
        buttonClass={styles.buttonLarge}
      />
      {lastSearches.map(PreviousSearchButton)}
      <List list={stories.data} dispatchStories={dispatchStories}/> 
    </div>
  )
}



export default App

export { storiesReducer }


