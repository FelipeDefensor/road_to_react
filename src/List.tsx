import * as React from 'react'
import { sortBy } from 'lodash'

import styles from "./App.module.css"
import { Story, Action, ReducerAction } from "./App"


export enum SORT_FIELD {
  TITLE = 'title',
  AUTHOR = 'author',
  COMMENTS = 'num_comments',
  POINTS = 'points',
  NONE = 'none',
}

const SORTS = {
  'none': (list: Story[]) => list,
  'title': (list: Story[]) => sortBy(list, 'title'),
  'author': (list: Story[]) => sortBy(list, 'author'),
  'num_comments': (list: Story[]) => sortBy(list, 'num_comments').reverse(),
  'points': (list: Story[]) => sortBy(list, 'points').reverse(),
};

const List = ({ list, dispatchStories }: {
  list: Story[],
  dispatchStories: (action: ReducerAction) => void 
  }) =>  {

  console.log('Rendering List')

  const [sort, setSort] = React.useState<keyof typeof SORTS>('none');
  const [isSortReversed, setReverseSort] = React.useState(false)  

  const handleRemoveItem = (story: Story ) => {
    dispatchStories({type: Action.REMOVE_STORY, payload: story})
  }
  
  const handleSortChange = (field: SORT_FIELD) => {
      if (field === sort) {
        setReverseSort(!isSortReversed)
      } else {
        setReverseSort(false)
        setSort(field)
      }
      console.log(isSortReversed)
    }
  

  function itemFromStory (story: Story) {
    return (   
    <Item key={story.objectID} story={story} onRemoveItem={handleRemoveItem}/>
  )}

  const sortFunction = SORTS[sort];
  const sortedList = isSortReversed ? sortFunction(list).reverse() : sortFunction(list);

  const getReversedIndicator = (reverse: Boolean) => {
    return !reverse ? '↑' : '↓'
  }

  return ( 
    <>
    <ul>
      <li style={{ display: 'flex' }}>
        <span style={{ width: '40%'}}>
          <button onClick={() => handleSortChange(SORT_FIELD.TITLE)} className={styles.sortHeader}>
            Title {sort === SORT_FIELD.TITLE && getReversedIndicator(isSortReversed)}
          </button>
        </span>
        <span style={{ width: '30%'}}>
          <button onClick={() => handleSortChange(SORT_FIELD.AUTHOR)} className={styles.sortHeader}>
            Author {sort === SORT_FIELD.AUTHOR && getReversedIndicator(isSortReversed)}
          </button>
        </span>
        <span style={{ width: '10%'}}>
          <button onClick={() => handleSortChange(SORT_FIELD.COMMENTS)} className={styles.sortHeader}>
            Comments {sort === SORT_FIELD.COMMENTS && getReversedIndicator(!isSortReversed)}
          </button>
        </span>
        <span style={{ width: '10%'}}>
          <button onClick={() => handleSortChange(SORT_FIELD.POINTS)} className={styles.sortHeader}>
            Points {sort === SORT_FIELD.POINTS && getReversedIndicator(!isSortReversed)}
          </button>
        </span>
        <span style={{ width: '10%'}}>
          Actions
        </span>
      </li>      
      {sortedList.map(itemFromStory)}
    </ul>
    </>
  )
  }


const Item = ({
story,
onRemoveItem
}: { 
story: Story,
onRemoveItem: (story: Story) => void
}): JSX.Element => {

const { title, url, author, num_comments, points } = story

return (
  <li className={styles.item}>
    <span style={{ width: '40%'}}>
      <a href={url}>{title}</a>
    </span>
    <span style={{ width: '30%'}}>Author: {author}</span>
    <span style={{ width: '10%'}}>Comments: {num_comments}</span>
    <span style={{ width: '10%'}}>Points: {points}</span>
    <span style={{ width: '10%'}}>
      <button onClick={() => onRemoveItem(story)} className={`${styles.button} ${styles.buttonSmall}`}>Dismiss</button>
    </span>
  </li>
  )
}


export { List, Item }
  
  