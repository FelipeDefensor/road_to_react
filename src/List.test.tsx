import { describe, it, expect, vi } from 'vitest'
import {
    render,
    screen,
    fireEvent,
} from '@testing-library/react'

import { List, Item } from './List'
import { SearchForm } from './SearchForm'
import { storiesReducer } from './App'
import { InputWithLabel } from './InputWithLabel'
import { scryRenderedComponentsWithType } from 'react-dom/test-utils'

const storyOne = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
}

const storyTwo = {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andre Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
}

const stories = [storyOne, storyTwo]

describe('Item', () => {
    it('renders all properties', () => {
        render(<Item story={storyOne} onRemoveItem={() => {}}/>);
  
        // expect(screen.getByText('Jordan')).toBeInTheDocument(); // couldn't get this to work
        expect(screen.getByText('React')).toHaveAttribute(
            'href',
            'https://reactjs.org/'
        );
    });

    it('renders a clickable dismiss buton', () => {
        render(<Item story={storyOne} onRemoveItem={() => {}}/>)

        expect(screen.getByRole('button')).toBeInTheDocument();      
    })

    it('clicking the dismiss button calls the callback handler', () => {
        const handleRemoveItem = vi.fn()

        render(<Item story={storyOne} onRemoveItem={handleRemoveItem} />)

        fireEvent.click(screen.getByRole('button'))

        expect(handleRemoveItem).toHaveBeenCalledTimes(1)

    })

    it('matches snapshot', () => {
        const { container } = render(<Item story={storyOne} onRemoveItem={() => {}}/>)
        expect(container.firstChild).toMatchSnapshot()
    })


});


describe('List', () => {

    it('displays stories title', () => {
        render(<List list={stories} dispatchStories={vi.fn()}/>)

        expect(screen.getByText(/Redux/)).toBeInTheDocument()
    })

    it('displays stories author', () => {
        render(<List list={stories} dispatchStories={vi.fn()}/>)

        expect(screen.getByText(/Dan Abramov/)).toBeInTheDocument()
    })

    it('removes story when remove button is clicked', () => {
        const removeStory = vi.fn()
        render(<List list={[storyOne]} dispatchStories={removeStory} />)
        
        fireEvent.click(screen.getByText('Dismiss'))

        expect(removeStory).toHaveBeenCalledTimes(1)

    })

    it('matches snapshot', () => {
        const { container } = render(< List list={stories} dispatchStories={() => {}}/>)
        expect(container.firstChild).toMatchSnapshot()
    })
})



