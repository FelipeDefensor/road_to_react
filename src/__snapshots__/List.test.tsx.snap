// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item > matches snapshot 1`] = `
<li
  class="_item_6dc87e"
>
  <span
    style="width: 40%;"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    style="width: 30%;"
  >
    Author: 
    Jordan Walke
  </span>
  <span
    style="width: 10%;"
  >
    Comments: 
    3
  </span>
  <span
    style="width: 10%;"
  >
    Points: 
    4
  </span>
  <span
    style="width: 10%;"
  >
    <button
      class="_button_6dc87e _buttonSmall_6dc87e"
    >
      Dismiss
    </button>
  </span>
</li>
`;

exports[`List > matches snapshot 1`] = `
<ul>
  <li
    style="display: flex;"
  >
    <span
      style="width: 40%;"
    >
      <button
        class="_sortHeader_6dc87e"
      >
        Title
      </button>
    </span>
    <span
      style="width: 30%;"
    >
      <button
        class="_sortHeader_6dc87e"
      >
        Author
      </button>
    </span>
    <span
      style="width: 10%;"
    >
      <button
        class="_sortHeader_6dc87e"
      >
        Comments
      </button>
    </span>
    <span
      style="width: 10%;"
    >
      <button
        class="_sortHeader_6dc87e"
      >
        Points
      </button>
    </span>
    <span
      class="_sortHeader_6dc87e"
      style="width: 10%;"
    >
      Actions
    </span>
  </li>
  <li
    class="_item_6dc87e"
  >
    <span
      style="width: 40%;"
    >
      <a
        href="https://reactjs.org/"
      >
        React
      </a>
    </span>
    <span
      style="width: 30%;"
    >
      Author: 
      Jordan Walke
    </span>
    <span
      style="width: 10%;"
    >
      Comments: 
      3
    </span>
    <span
      style="width: 10%;"
    >
      Points: 
      4
    </span>
    <span
      style="width: 10%;"
    >
      <button
        class="_button_6dc87e _buttonSmall_6dc87e"
      >
        Dismiss
      </button>
    </span>
  </li>
  <li
    class="_item_6dc87e"
  >
    <span
      style="width: 40%;"
    >
      <a
        href="https://redux.js.org/"
      >
        Redux
      </a>
    </span>
    <span
      style="width: 30%;"
    >
      Author: 
      Dan Abramov, Andre Clark
    </span>
    <span
      style="width: 10%;"
    >
      Comments: 
      2
    </span>
    <span
      style="width: 10%;"
    >
      Points: 
      5
    </span>
    <span
      style="width: 10%;"
    >
      <button
        class="_button_6dc87e _buttonSmall_6dc87e"
      >
        Dismiss
      </button>
    </span>
  </li>
</ul>
`;
